### PR TITLE
build job names dynamically of component, pipelinename, timestamp

### PR DESCRIPTION
**Pull request by benkoller**
This is a quick attempt to mitigate #410 in a minimally-invasive fashion.

### CODE REVIEW

**Comment by benkoller** [comment]
@googlebot I signed it!

File: tfx/orchestration/beam/beam_dag_runner.py [hunk 1]
```diff
@@ -93,6 +106,22 @@ def run(self, tfx_pipeline: pipeline.Pipeline) -> None:
...
+      '{pipeline_name}-{ts}'.format(
+        pipeline_name=tfx_pipeline.pipeline_info.pipeline_name,
+        ts=int(datetime.datetime.timestamp(datetime.datetime.now()))
+      )
+    )
+    # Exploit the last-in priority of --job_name for beam
+    if self._beam_orchestrator_args:
+      self._beam_orchestrator_args.append('--job_name={}'.format(job_name))
+    else:
+      self._beam_orchestrator_args = ['--job_name={}'.format(job_name)]
```
Comment (1025KB) [inline tfx/orchestration/beam/beam_dag_runner.py]: the beam orchestrator is for entire pipeline instead of a single component

Comment (benkoller) [reply tfx/orchestration/beam/beam_dag_runner.py]: Hi @1025KB , maybe I have a misunderstanding then and you can help me by clarifying. I was under the assumption from the comments in the code that `self._beam_orchestrator_args` are responsible for the orchestrator pipeline args and that `component_launcher.ComponentLauncher()` takes `additional_pipeline_args`, which in turn contain a list of `beam_pipeline_args`. This drove me to these design decisions.

Please also note that the reviewed code contained a bug under which the actual pipeline names for executors could be wrong due to a race-condition-like flaw in my own naming logic. According to my tests its fixed now.

**Comment by benkoller** [comment]
Cool that 0.14. is out now, is there a way I can assist in putting this onto the roadmap for 0.14.1 / 0.15.0?

**Comment by 1025KB** [comment]
Hi, if you are trying to run tfx on dataflow, could you try kubeflow_dag_runner? which I think already handles those problems

**Comment by benkoller** [comment]
Hi @neuromage thanks for taking a look.

Dataflow creates pipeline names by itself, in the format `beam-<USERNAME>-<timestamp`. This creates opaque pipeline names for both the orchestrator as well as the components.

By creating pipeline names based on a more descriptive pattern in accordance to the tfx `pipeline_name` makes individual runs identifiable. Especially in an environment of multiple parallel experiments and a CI/CD approach this is import for us to correlate code, config and experiment.

Keeping our use case aside it seems counter-intuitive to specify a `pipeline_name` and end up with Dataflow pipelines that can't easily be correlated via name. This fixes this.

In a more elaborate solution the beam arguments for orchestrator and components would be passed not as a list of strings but as a dictionary that is then deconstructed into the Dataflow/Beam command line parameters downstream. My fix is really just a simple fix in this comparison.

**Comment by neuromage** [comment]
Thanks @benkoller. I think ideally, we'd like a solution that works across orchestrators, and not just for the BeamDagRunner. Given this, I suggest we close this PR and wait for the more general solution. Does that work for you?

**Comment by benkoller** [comment]
@neuromage if a generalised solution is possible that would definitely work for me. Out of curiosity, at what point in the code would you handle the job namings? Or would you still handle namings in the orchestrator runners, but via a util?

**Comment by neuromage** [comment]
I haven't thought about this enough, but off the top of my head, for individual Dataflow jobs, I would most likely update this using a combination of `beam_pipeline_args` and executor `context`:
https://github.com/tensorflow/tfx/blob/d651dfda08278dd46a32b56931c6a61d7ed2e3c1/tfx/components/base/base_executor.py#L34


Thanks again for the PR! I'm going to go ahead and close this one for now.

**Pull request closed without merging**